### ISSUE TITLE
errorrtrace uses very large amounts of memory on Racket CS (and runs out of memory)

### ISSUE THREAD

**Issue by alex-hhh** [racket/racket#3459]
**What version of Racket are you using?**

7.8.0.900 CS, 64 bit version, on a Windows platform

**What program did you run?**
Please include a short example that triggers the bug

https://github.com/alex-hhh/ActivityLog2, without any complied ZO files (except for dependent libraries), run errrortrace like so:

```
racket -l errortrace -t run.rkt
```

**What should have happened?**

Application should have started up

**If you got an error message, please include it here.**

Memory use grew to about 12 Gb and the application exited with an "out of memory" message.  My machine has 16 Gb of RAM.

**Please include any other relevant details**
e.g., the operating system used or how you are running the code.

Running the same command in Racket 7.8 BC uses about 0.7 Gb of RAM and the application starts up.  Without errortrace instrumentation, the application uses about 0.6 Gb or RAM.

**Comment by pmatos** [comment]
Might be Windows related. Compiled master and I have the same results on CS (Linux) as you report for Racket 7.8 BC: < 1Gb RAM used, with errortrace consuming only a bit more memory.

**Comment by mflatt** [comment]
I haven't been able to replicate the problem on Windows, either with the 7.8.0.900 CS pre-release or a slightly newer HEAD build.

Does running as
```
racket -l mzlib/traceld -l errortrace -t run.rkt
```
provide any hints? Maybe that will show a particular file where loading goes wrong.

**Comment by alex-hhh** [comment]
It looks like the problem is intermittent on my machine as well.  I could never reproduce it when using the mzlib/traceld library:

```
racket -l mzlib/traceld -l errortrace -t run.rkt
```

but I am able to reproduce the problem using, although less often, perhaps 10-15% of the time (although sometimes it happens several times in a row).  I could not figure out what triggers it, as I don't make any changes to my machine or the source code between runs.

```
racket -l errortrace -t run.rkt
```



<img width="656" alt="high-memory-usage" src="https://user-images.githubusercontent.com/11592690/97062745-d96a4a00-15ce-11eb-8030-c5c462b84b84.png">

**Comment by alex-hhh** [comment]
I managed to reproduce this problem using the traceld library.  The log stops printing after loading the dbsystem_rkt.zo library, but the memory usage continues to grow and the application is using 15-20% CPU, until it eventually runs out of memory.  On a good run, the following libraries are loaded after dbsystem_rkt.zo: sql-data_rkt.zo, win32_rkt.zo and add-cert_rkt.zo.  I attached a good run log file (i.e. when the application loaded fine) as well as the last entries from the bad run.

One which may be relevant is that I changed the application to use a separate place for the sqlite database connection (this is done by passing `#:use-place #t` to `sqlite3-connect`).  The application seems to hang right after loading the libraries for the place.   This is a recent change (September 2020) and, so far, I have only used it with Racket BC, but without any problems.  I only downloaded 7.8.0.900 a few days ago and did not work on ActivityLog2 too much (perhaps this has nothing to do with errortrace, but some problem in the places code?)

[bad-run-fragment.txt](https://github.com/racket/racket/files/5432779/bad-run-fragment.txt)

[good-run.txt](https://github.com/racket/racket/files/5432783/good-run.txt)

**Comment by mflatt** [comment]
Thanks! This is all helpful. I still haven't replicated the problem in 12 runs, so any detail might help.

Places might indeed be important. Yesterday, I found and fixed a couple of CS problems with places — but the problems were with subprocesses and IP address resolution, neither of which look relevant here.

I don't suppose you have more of the bad-run trace? I notice that the bad run has already taken twice as long as the good run by the time it reaches "place.rkt", and it might be useful to know whether the slowdown is uniform or whether there was a delay in some specific point.

**Comment by alex-hhh** [comment]
Unfortunately I don't have any more of the "bad run" log, as I didn't really expect to experience the problem again,  and I was using the error trace library to debug a problem with ActivityLog2 itself (unrelated to this issue, but unfortunately also difficult to reproduce).

With regards to timings, the main difference between the runs is that the bad run was done on battery power only (where my laptop runs at a slower speed), while the good run was taken with the laptop connected to the main power supply.  I recorded a good run on battery power and the times seem closer to the "bad run", although they are still faster...

As for how to reproduce it, after experiencing that bad run, I tried to reproduce it 17 times and they all started fine, so I am not really sure what triggers the actual problem.  I also tried to write a simple program which just opens a SQLite database using places and runs a simple query, but that program also runs fine, without any problems.

[good-run-battery-power.txt](https://github.com/racket/racket/files/5432984/good-run-battery-power.txt)

**Comment by mflatt** [comment]
Ok, that makes sense - thanks!

**Comment by alex-hhh** [comment]
This is a full log for a "bad run" where it runs out of memory.  The log was recorded while the laptop was powered from the main power supply (so it was running at maximum capacity).  Also, some of the packages (geoid and gui-widget-mixins) are loaded from a different location -- I experience this problem while working on the application itseld.

Also note that the last log entry is at about 100 seconds into the startup, but it took more than 5 minutes to actually crash (when the "out of memory" message is printed.

[bad-run.txt](https://github.com/racket/racket/files/5436817/bad-run.txt)

**Comment by mflatt** [comment]
I'm still having trouble replicating this problem. My best guess is that it's the timing on my Windows VM, or something like that.

Could it have anything to do with existing database state? That is, do you see the problem even starting with clean database state, or is there something I should do to ensure that before a run?

**Comment by alex-hhh** [comment]
Unfortunately, I have never been able to reproduce this when I actively tried, so I cannot really test any scenarios.  So far this is what I was able to determine:

* at least once, the problem happened without running errortrace, simply by running the application under racket
* every time it happened it happened on the production database
* every time it happened it happened at the start of the program (i.e. just after the database connection was made and perhaps some of the initial SQL queries were run).  It never happened during running the program, when more data is fetched from the database
* I tried to write a simple program which opens the production database using a separate place and run a query on it (trying to isolate the problem to just opening the database), but this program did not have any problems

What I plan to do next is to update to the newly released 7.9 CS (I was using a pre-release version which is now a few weeks old), and, when (if?) the problem happens, I will run procmon and vmmap to try to figure out what is going on (it takes about 5 minutes to run out of memory).

**Comment by mflatt** [comment]
FWIW, there are already several improvements and repairs to CS relative to v7.9. Some of them are related to places — but they'reni the interaction of places and subprocesses and network-address resolution, which doesn't seem relevant here. So, my best guess is that v7.9.0.3 will behave the same as v7.9 and the pre-release version.

**Comment by alex-hhh** [comment]
The problem happened twice in a row, and I managed to collect some more data (see link below).  Both runs were without error trace, i.e. just "racket run.rkt".

In the first case, I started procmon to monitor system calls for a few seconds while the application was running.   You can find the logs in the procmon-log.zip file.   Unfortunately, nothing interesting is in there.  Even though the application consumed about 40% of the CPU and kept increasing its memory, there are no mmap system calls to request more memory from the system.

Also for the first case, I collected a memory snapshot using vmmap, this file is in vmap-dump.zip.

For the second run, I attached the Visual Studio debugger and collected two dump files, a short interval one after the other.  These are the Racket.zip and Racket 2.zip files (the dump files itself are 5 and 6 Gb in size).  I am hoping that we can somehow compare the two dump files to determine the contents of the new memory that was allocated in between the memory dumps.  Unfortunately, my skill in this area is limited.

I have also added the exact Racket installer that I used, in case the binaries are needed for the dump files.

https://1drv.ms/u/s!AlQQEXklnD7vhfJrdT_n-pbx6RBmxQ?e=SFndHv

**Comment by alex-hhh** [comment]
For what is worth, here are the call stacks for the two active threads in the first dump files.  Since I don't have the symbols, everything but the first few entries, are probably incorrect (e.g. frame 0xa is probably not in `Scompact_heap`.  The call stacks for the second dump file look the same.

Is there a possibility to obtain the symbol files for the build I was using?

```
0:000> ~0kn
 # Child-SP          RetAddr           Call Site
00 00000000`0061f958 00007ffd`ff44a443 win32u!NtUserPeekMessage+0x14
01 00000000`0061f960 00007ffd`ff44a3a3 user32!_PeekMessage+0x43
02 00000000`0061f9d0 00000000`093ebc14 user32!PeekMessageW+0x143
03 00000000`0061fa40 00000001`470beda8 0x93ebc14
04 00000000`0061fa48 00000000`2966c400 0x00000001`470beda8
05 00000000`0061fa50 00000000`2966c400 0x2966c400
06 00000000`0061fa58 00000000`0ac07ce0 0x2966c400
07 00000000`0061fa60 00000000`00000001 0xac07ce0
08 00000000`0061fa68 00000000`06b1ef62 0x1
*** ERROR: Symbol file could not be found.  Defaulted to export symbols for libracketcs_bpiwzo.dll - 
09 00000000`0061fa70 00000000`6a37cf80 0x6b1ef62
0a 00000000`0061fa78 00000000`6a18e54b libracketcs_bpiwzo!Scompact_heap+0x1e53f0
0b 00000000`0061fa80 00000000`6a18e733 libracketcs_bpiwzo!Sput_arg+0x46b
0c 00000000`0061fbc0 00000000`6a182772 libracketcs_bpiwzo!Scall2+0x53
*** WARNING: Unable to verify timestamp for Racket.exe
*** ERROR: Module load completed but symbols could not be loaded for Racket.exe
0d 00000000`0061fc10 00000000`00402885 libracketcs_bpiwzo!racket_boot+0x352
0e 00000000`0061fcb0 00000000`004013c7 Racket+0x2885
0f 00000000`0061fe30 00000000`004014fb Racket+0x13c7
10 00000000`0061ff00 00007ffd`ff667034 Racket+0x14fb
11 00000000`0061ff30 00007ffe`00b7cec1 kernel32!BaseThreadInitThunk+0x14
12 00000000`0061ff60 00000000`00000000 ntdll!RtlUserThreadStart+0x21

0:000> ~3kn
 # Child-SP          RetAddr           Call Site
00 00000000`22fcfba8 00007ffe`00b952e5 ntdll!NtWaitForAlertByThreadId+0x14
01 00000000`22fcfbb0 00007ffd`fe44aab9 ntdll!RtlSleepConditionVariableCS+0x105
02 00000000`22fcfc20 00000000`6a18fa52 KERNELBASE!SleepConditionVariableCS+0x29
03 00000000`22fcfc50 00000000`068c6fc7 libracketcs_bpiwzo!Sdestroy_thread+0x492
04 00000000`22fcfce0 00000000`05379b50 0x68c6fc7
05 00000000`22fcfce8 00000000`1cce41b0 0x5379b50
06 00000000`22fcfcf0 00000000`0bfc3b90 0x1cce41b0
07 00000000`22fcfcf8 00000000`052ccfc0 0xbfc3b90
08 00000000`22fcfd00 00000000`05378aa0 0x52ccfc0
09 00000000`22fcfd08 00000000`6a18e54b 0x5378aa0
0a 00000000`22fcfd10 00000000`6a18e5c1 libracketcs_bpiwzo!Sput_arg+0x46b
0b 00000000`22fcfe50 00000000`6a18f6e8 libracketcs_bpiwzo!Scall0+0x41
0c 00000000`22fcfe90 00007ffd`ff7fad42 libracketcs_bpiwzo!Sdestroy_thread+0x128
0d 00000000`22fcfed0 00007ffd`ff7fae1c msvcrt!_callthreadstart+0x1e
0e 00000000`22fcff00 00007ffd`ff667034 msvcrt!_threadstart+0x7c
0f 00000000`22fcff30 00007ffe`00b7cec1 kernel32!BaseThreadInitThunk+0x14
10 00000000`22fcff60 00000000`00000000 ntdll!RtlUserThreadStart+0x21
```

**Comment by mflatt** [comment]
That stack trace seems about as helpful as we're likely to get, even with more symbols, since it looks like we're into Racket-generated code after a few frames (i.e., the C stack can't actually say much).

Just to make sure: You're not able to hit Ctl-C and get any kind of Racket stack trace when the program is starting to use too much memory?

The main thing the stack trace suggests for me is that the out-of-memory problem is not happening during GC, but instead through something that involves handling Windows messages. In other words, while nothing's definite, I'm starting to think that the problem isn't in the core implementation of places/threads in CS, but something wrong in Racket libraries — and something that happens when timing lines up a certain way.

I still can't point to anything new since v7.8.0.900 that is obviously relevant, there are improvements that could be relevant. It would be good to confirm that the problem still happens with v7.9.0.4, just so I can stop wondering if any of the repairs are relevant. :)

**Comment by alex-hhh** [comment]
Unfortunately, it never occured to me to stop the program with Ctrl+C, I always stopped either by clicking on the close button on the splash screen window, or killing it from the process monitor.

I have now installed 7.9.0.4 and will start using it.  It might take several days until I notice the problem again (unless it was already fixed).

**Comment by alex-hhh** [comment]
Unfortunately it happened again with v7.9.0.4 [cs].  I was running the "racket run.rkt" command from an Emacs shell buffer and, at least on Windows, it does not work to sent a process a BREAK signal from Emacs, so I could not get a Racket stack trace.

**Comment by mflatt** [comment]
The next snapshot (20201119 or later) will include a repair that is possibly relevant, since there was a problem with thread termination and `sync` or `thread-dead?` on a thread terminated thread if the scheduler's timer expired at just the wrong time. 

More likely, a debugging environment variable associated with the change could help: Try setting `PLT_THREAD_QUANTUM` to something much smaller than 100000. A value of 1000 would make the Racket thread scheduler swap threads 100 times as frequently as usual, which can help in exposing a race condition. Maybe some value of `PLT_THREAD_QUANTUM` will make the problem happen more on your machine.

I still wasn't able to replicate the problem in my installation when setting `PLT_THREAD_QUANTUM`. So, maybe it won't help, or maybe I'm missing some other ingredient in my installation.

**Comment by alex-hhh** [comment]
This morning, I installed the latest snapshot (20210101-db9c06a, version 7.9.0.22 CS) and this problem is still happening.   

Setting PLT_THREAD_QUANTUM to 1000 does not make it easier to reprotduce it though...

**Comment by mflatt** [comment]
I'm finally able to provoke bad behavior! it seems to be related to the splash screen.

If I change "﻿plt:ActivityLog2-splash-max-width" in "racket-prefs.rktd" to a smaller number, then startup seems to go bad at the place where the splash screen would go away. At that point, the splash screen would be trying to update the preferencs file with a new number, and probably something goes wrong updating the file. It also seems likely that you didn't intend for a Racket preferences file to get written/update, in any case.

**Comment by mflatt** [comment]
I haven't been able to provoke the problem a second time, so just changing splash-max-width is not enough by itself. But maybe it's an ingredient in replicating the problem.

**Comment by alex-hhh** [comment]
Its really good that you managed to replicate the problem, I was starting to worry that there was something specific to my machine.  

The splash screen used by ActivityLog2 is the same as the one used by DrRacket, but I never seen DrRacket experience this problem at startup.  Also, the splash screen is explicitly closed when the toplevel window is created (in "rkt/toplevel.rkt"), by calling `shutdown-splash` and `close-splash`, it does not rely on the number of items loaded (as far as I understand it).

**Comment by mflatt** [comment]
I mean that the framework splash screen sets the load handler to count how many files are loaded, and then `close-splash` records in the Racket preferences file number of loaded file. You've turned off the progress bar for ActivityLog2, though, so that count isn't actually used.

Unfortunately, I still haven't been able to provoke the problem a second time, but it's good to have evidence that it really isn't specific to your machine.

**Comment by mflatt** [comment]
I found a way to provoke bad behavior: change the preferences file to have the wrong count for `plt:ActivityLog2-splash-max-width`, then use `type racket-prefs.rktd | more` and not advance to effectively make the file locked. Then, when ActivityLog2 gets to the point where it would update the preferences file, it starts using memory very quickly.  I can stop it by clicking the splash screen's close box.

If you try something like that, does it look the way that other bad cases look? When the problem has happened in the past, was DrRacket running (or even some other program that might have touched the preferences file) at the same time? Even if not, it's at least possible that a background Windows process was scanning the file.

I haven't yet investigated why attempting to write to a preferences file locked by `type` goes so wrong.

**Comment by alex-hhh** [comment]
I can reproduce the problem following your steps, but also when the preferences file is not locked and no other racket process is running.  It seems to me that setting `plt:ActivityLog2-splash-max-width` to 1 appears to trigger the problem and setting this value to 1000, which still requires updating the preferences file, does not trigger the problem.  I tried this a few times and seemed consistent.

Also, when compiling the Racket files, the resulting run has no trouble updating the value 1 with the actual number of modules loaded...

I should also note that the `plt:ActivityLog2-splash-max-width` is set to 1 quite frequently on my machine, as this is the value stored by a built executable (presumably because no external modules are loaded).  I keep a built and installed version of ActivityLog2 on my machine to actually analize training data, and whenever I start that build, the value gets written to 1.

**Comment by mflatt** [comment]
Setting the width to 1 usually works for me, too, though somewhat less reliably. In any case, these adjustments appear to be just a way to make timing or allocation trigger the real problem. I'm able to replicate the problem even when running compiled files, which helps a lot.

So far, the problem appears to be in the place channels, as you suspected back in October. I'll continue to debug.

**Comment by mflatt** [comment]
The problem does not appear to be place channels after all. Instead, it looks like this is a bug in `racket/gui` — one that's not CS-specific.

Place channels getting stuck was a symptom of a place generally getting stuck, which happened at the point where the place requested a GC. The GC never happened because interrupts were disabled in the main place. (That also explain why memory use increases rapidly: GCs were disabled in general.) 

Interrupts were left in a disabled state, meanwhile, because they were left *enabled* on entry to foreign-function callback.
I am not entirely clear on how interrupts were enabled on entry to a callback, but it seems like they were left enabled in a callback because an earlier callback that was supposed to be atomic wasn't declared atomic. That callback is the general event-handling callback attached to dialogs, and the repair is racket/gui@9f32493f7ef8ff39282f6f2be1c773ef729a6d99. (This non-atomic-callback bug could cause any amount of trouble for GUI programs on Windows, so it was well worth tracking down, whatever else results here.)

Not being able to replicate the problem anymore is weak evidence of a repair, of course, given how fragile the bad state was. But I have some hope that the `racket/gui` repair really fixes the problem. It's a clear bug, and I found it through a chain of effects in failing runs here.

There's a question of whether non-atomic callbacks should be allowed at all. They usually create problems, and maybe they always do. I'll look into that more.

`PLT_THREAD_QUANTUM` turned out to be a key tool for tracking down the problem, since it was related to scheduling. Small changes to the program would cause the crash to go away, and I had to tweak `PLT_THREAD_QUANTUM` each time (by trial and error) to find a new crashing value after applying any sort of debugging instrumentation.

**Comment by alex-hhh** [comment]
Unfortunately, I had this problem happen again with Racket CS 7.9.0.900 downloaded today from https://pre-release.racket-lang.org/.  I am not sure how to check if this build has the fix or not (I assume it does), but the SHA1 sum is 6da3b0b47d117c4975aa11959cb3f539ef000778.

I also had problems with a previous 7.9.0.900 install (from around January 18), but I wanted to wait a bit before reporting it as I wasn't sure that the fix made into the build.  Also with the previous 7.9.0.900 install, I noticed the high memory use at application shutdown.  I.e. I closed the main window and the application remained running without a window and memory use was increased.  This only happened once.

**Comment by mflatt** [comment]
Over the past few days, I have been trying to set up a more automated way of triggering the problem, and I haven't yet succeeded — even after rewinding the previous repair, so that I know at least one kind of bug to try to provoke. Provoking the bug seems to require me interacting with the machine, maybe by moving the mouse to generate events for the splash screen.

My next idea is to try some GUI-testing tool that generates Win32-level events for the window, and maybe I'll be able to automate triggering bad behavior that way.

**Comment by mflatt** [comment]
Ryan noticed a bug that could be related where interrupts were meant to be disabled but left enabled, fixed in 920e3dbde5954c2858129d04aa6207a43ff16f43 (should be in snapshots dated 20210329 and later).

**Comment by alex-hhh** [comment]
I am sorry to report that this problem is still happening with a build from March 31st, which I believe contains commit https://github.com/racket/racket/commit/920e3dbde5954c2858129d04aa6207a43ff16f43

```
Welcome to DrRacket, version 8.0.0.12--2021-03-31(9a4f968/a) [cs].
Language: racket/gui [custom].
```

**Comment by mflatt** [comment]
Yet another repair that is potentially related: 6d30ff78c14072ce5388a659a9d3415d2249d222

**Comment by alex-hhh** [comment]
Unfortunately, the problem is still happening in version 8.1.0.2--2021-04-20(cc0b1ba/a) [cs].

**Comment by mflatt** [comment]
Ryan made several changes to the way the SQLite3 binding uses places, and `#:use-place #t` now uses Chez-level threads (not Racket places). It would be good to know if you still see problems.

There have been two other repairs to fix problems that had similar symptoms, but they don't seem likely to be related. One was a problem with `shell-execute`, and another was a problem with glib logging in a foreign thread.

**Comment by alex-hhh** [comment]
I have been using 8.1.0.6--2021-06-09(97e96cd/a) since Friday and haven't noticed any problems so far.

**Comment by alex-hhh** [comment]
Hi @mflatt , I have been using 8.1.0.6 for the past month and recently updated to 8.2 and have not noticed this issue anymore.  I think this can be closed now.

**Comment by mflatt** [comment]
Whew! Thanks, @alex-hhh.

**Comment by alex-hhh** [comment]
I am sorry to say that this problem is still happening, and recently I discovered a new scenario in which it is triggered: the application has a plot where a large number of points (2 million) are displayed on a map, in an interactive way, streaming data from the SQLite database, grouping the data, showing it on the map, and allowing the user to interact with the map while the data is filled in.   Occasionally, when zooming in the map while data is streamed in, the application starts to allocate memory until the system kills it.  Unfortunately, this case is not easy to reproduce either.

I managed to trigger a memory dump while the application was allocating memory and also another memory dump from an application run where the problem did not happen.  Comparing the memory layout of the two memory dumps I could identify the "extra" memory being allocated and, inspecting it, it looks like most of it is pointers (the upper 32 bits are either 0 or 1 making each 64 bit value a valid address):

```
0:000> dq 1`2ec60000 L40800
dq 1`2ec60000 L40800
00000001`2ec60000  00000001`2ec43b8f 00000001`2ebcb229
00000001`2ec60010  00000000`07792320 00000001`2e4aebf0
00000001`2ec60020  00000000`00000018 00000000`00000018
00000001`2ec60030  00000000`074f3ffd 00000000`0b511c28
00000001`2ec60040  00000000`00000026 00000000`00000026
00000001`2ec60050  00000000`3309a2b0 00000000`09d5d88f
00000001`2ec60060  00000000`09d5d8bf 00000000`09d5d8af
00000001`2ec60070  00000000`3309a460 00000000`09d5d8bf
00000001`2ec60080  00000000`09d5d8af 00000000`0910d2ad
00000001`2ec60090  00000001`2ec43abd 00000000`09d5d879
00000001`2ec600a0  00000000`09d5d8cf 00000000`09d580fd
00000001`2ec600b0  00000000`3309ab70 00000000`09d5d88f
00000001`2ec600c0  00000000`085a6b1f 00000000`00000000
00000001`2ec600d0  00000001`2ec6000d 00000001`2ec43b79
(many lines omitted)
``` 

With some work, I think it is possible to identify what each data value represents and I could try to correlate it with the data the application should contain (since I am familiar with the later).   Would this information be useful in determining what the cause of the problem is?

If this is useful, I would need information about the tags that Chez uses for each data type, is this information available somewhere?

Also, this happens using Racket 8.4 CS on Windows.

Thanks,
Alex.
